src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -0,0 +1,526 @@
+		return
+	}
+}
+
+func (r *downloadRequest) createRemoteRequest() (*http.Response, *util.JSONResponse) {
+	urls := getMatrixUrls(r.MediaMetadata.Origin)
+
+	r.Logger.WithField("URL", urls[0]).Info("Connecting to remote")
+
+	remoteReqAddr := urls[0] + "/_matrix/media/v1/download/" + string(r.MediaMetadata.Origin) + "/" + string(r.MediaMetadata.MediaID)
+	remoteReq, err := http.NewRequest("GET", remoteReqAddr, nil)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.Unknown(fmt.Sprintf("File with media ID %q could not be downloaded from %q", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)),
+		}
+	}
+
+	remoteReq.Header.Set("Host", string(r.MediaMetadata.Origin))
+
+	client := http.Client{}
+	resp, err := client.Do(remoteReq)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: 502,
+			JSON: jsonerror.Unknown(fmt.Sprintf("File with media ID %q could not be downloaded from %q", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)),
+		}
+	}
+
+	if resp.StatusCode != 200 {
+		r.Logger.WithFields(log.Fields{
+			"Origin":     r.MediaMetadata.Origin,
+			"MediaID":    r.MediaMetadata.MediaID,
+			"StatusCode": resp.StatusCode,
+		}).Info("Received error response")
+		if resp.StatusCode == 404 {
+			r.Logger.WithFields(log.Fields{
+				"Origin":     r.MediaMetadata.Origin,
+				"MediaID":    r.MediaMetadata.MediaID,
+				"StatusCode": resp.StatusCode,
+			}).Warn("Remote server says file does not exist")
+			return nil, &util.JSONResponse{
+				Code: 404,
+				JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+			}
+		}
+		return nil, &util.JSONResponse{
+			Code: 502,
+			JSON: jsonerror.Unknown(fmt.Sprintf("File with media ID %q could not be downloaded from %q", r.MediaMetadata.MediaID, r.MediaMetadata.Origin)),
+		}
+	}
+
+	return resp, nil
+}
+
+func (r *downloadRequest) closeConnection(w http.ResponseWriter) {
+	r.Logger.WithFields(log.Fields{
+		"Origin":  r.MediaMetadata.Origin,
+		"MediaID": r.MediaMetadata.MediaID,
+	}).Info("Attempting to close the connection.")
+	hijacker, ok := w.(http.Hijacker)
+	if ok {
+		connection, _, hijackErr := hijacker.Hijack()
+		if hijackErr == nil {
+			r.Logger.WithFields(log.Fields{
+				"Origin":  r.MediaMetadata.Origin,
+				"MediaID": r.MediaMetadata.MediaID,
+			}).Info("Closing")
+			connection.Close()
+		} else {
+			r.Logger.WithError(hijackErr).WithFields(log.Fields{
+				"Origin":  r.MediaMetadata.Origin,
+				"MediaID": r.MediaMetadata.MediaID,
+			}).Warn("Error trying to hijack and close connection")
+		}
+	}
+}
+
+func completeRemoteRequest(activeRemoteRequests *types.ActiveRemoteRequests, mxcURL string) {
+	if activeRemoteRequestCondition, ok := activeRemoteRequests.Set[mxcURL]; ok {
+		activeRemoteRequestCondition.Broadcast()
+	}
+	delete(activeRemoteRequests.Set, mxcURL)
+	activeRemoteRequests.Unlock()
```

Comment:
Please don't do this. There's no documentation to say that calling this function will unlock the map, and mismatched locks/unlocks are a nightmare to debug. Keep locks/unlocks physically together in the code: make liberal use of `defer`. It exists for reasons like this.